README.md
# xlReg_go

The reg library for [xlattice_go.](https://jddixon.github.io/xlattice_go)

[xlReg](xlReg.html) is a tool, primarily intended for use in testing,
which facilitates the formation of
[clusters,](https://jddixon.github.io/xlCluster_go)
groups of cooperating
[nodes.](https://jddixon.github.io/xlNode_go)

On registration, a client (an xLattice cluster member)
is issued a globally unique NodeID, a 256-bit random value.
Once it has an ID, the member can create and/or join clusters.

A cluster has
a maximum size set when it is created.  When members join the cluster they
register their two RSA public keys and one or more IP addresses.
If the cluster only supports communications between members, members
register only one IP address.  If non-member clients are allowed to
communicate with the cluster, members register a second address for
that purpose.  It is possible that certain applications may require
additional IP addresses.  (The first address is used for communications
between cluster members.  Any second address is used for communications
between cluster members acting as servers and their clients.)

When a member has completed registration, it can retrieve
the configuration data other members have registered.

The xlReg server, its clients, and the cluster members, are all
[XLattice nodes](https://jddixon.github.io/xlattice_go/node.html)
which means that each has

* a unique [NodeID](https://jddixon.github.io/xlNodeID_go)
* two RSA keys, one for encrypting data and the other for creating
  digital signatures
* at least one address (such as a TCP/IP connection) on which it
  listens at all times
* and optionally a local file system

## Protocols

Cluster members communicate using a common
[protocol](https://jddixon.github.io/xlReg_go/xlReg_protocol.html)
based on Google's
[Protocol Buffers](https://jddixon.github.io/xlReg_go/xlReg_protobuf.html),
a widely-used standard.  The same protocol is used for communication with
the xlReg registry and for communications with other cluster members.
Communications sessions use AES keys that are unique to the session.

## Project Status

The Go version of the xlReg server has been stable and in production
use since May 2014, when it was split off from the xlattice_go project.

## On-line Documentation

More information on the **xlReg_go** project can be found
[here](https://jddixon.github.io/xlReg_go)


CHANGES
xlReg_go/CHANGES
-----------------------------------------------------------------------------
The version number and date in this file must be kept in sync with version.go
-----------------------------------------------------------------------------

v1.9.24
    2017-11-20
        * correct directory structure, config files                 SLOC 5046
v1.9.23
    2017-10-21
        * add projConfig.toml                                       SLOC 5087
v1.9.22
    2017-08-26
        * corrected path to log file                                SLOC 5066
v1.9.21
    2016-11-15
        * minor bug fixes: all tests succeed                        SLOC 5039
v1.9.20
    2016-09-21
        * standardize licensing                                     SLOC 5025
v1.9.19
    2016-07-05
        * add ghpDoc/{css,img}                                      SLOC 4993
v1.9.18
    2016-05-20
        * add bkp2U, .dvcz, etc                                     SLOC 5100
v1.9.17
    2016-02-27
        * update LICENSE                                            SLOC 4048
v1.9.16
    2016-01-28
        * minor edits                                               SLOC 4048
v1.9.15
    2016-01-27
        * minor edits                                               SLOC 4048
v1.9.14
    2016-01-20
        * protocol, protobuf pages linked to {README,index}
v1.9.13
    2016-01-14
        * drop use of code.google.com                               SLOC 3995
v1.9.12
    2015-12-25
        * fix dead link
v1.9.11
    2015-11-26
        * dropped duplicate *.md ; minor edits to all *.md          SLOC 3995
v1.9.10
    2015-10-02
        * added missing *.md to project directory                   SLOC 3995
        * edited all *.md
v1.9.9
    2015-06-10
        * changes reflecting API changst elsewhere                  SLOC 3971
v1.9.8
    2015-06-06
        * changed signature for NextNBLine                          SLOC 3972
v1.9.7
    2015-06-01
        * add scripts to daemonize xlReg on the server              SLOC 3941
v1.9.6
    2015-04-28
        * ATTR* constants moved to xlCluster                        SLOC 3918
v1.9.5
    2015-04-26
        * tests succeed after changes to aes_cnx API                SLOC 3924
v1.9.4
    2015-04-25
        * changes to matching refactoring in aes_cnx                SLOC 3937
v1.9.3
    2015-04-24
        * aes_cnx{_test}.go renamed cnxHandler{_text}.go
        * then restructured to use xlProtocol_go/aes_cnx properly   SLOC 3943
v1.9.2
    2015-04-23
        * working, tested aes_cnx_test.go                           SLOC 3998
v1.9.1
    2015-04-22
        * skeletal aes_cnx_test.go                                  SLOC 3998
v1.9.0
    2015-04-17
        * change handling of IV to follow xlProtocol/aes_cnx        SLOC 3924
v1.8.4
    2015-04-10
        * add -g globalAddr option, IP address to appear in 
            regCred.dat
v1.8.3
    2015-04-09
        * add portlangRegCred_test.go                               SLOC 4124
v1.8.2
    2014-12-27
        * correct bug in xlReg.go (OpenAcc() not called)            SLOC 4055
v1.8.1
    2014-12-17
        * import SHA*_BIN_LEN from xlUtil_go                        SLOC 4045
v1.8.0
    2014-12-02
        * rename Eph/RegServer.Close() to Stop()                    SLOC 4044
v1.7.1
    2014-11-26
        * added bootstrap.go from paxos_go/pktComms                 SLOC 4044
v1.7.0
    2014-11-24
        * improved handling of open/close acceptors
        * fix various bugs                                          SLOC 3957
v1.6.5
    2014-11-18
        * split out xlCluster_go as separarate project
        * cleaned up error codes                                    SLOC 3920
v1.6.4
    2014-11-14
        * improvements to cluster_member_test.go                    SLOC 4715
v1.6.3
    2014-11-13
        * added TestCluster                                         SLOC 4645
    2014-11-10
        * RegCluster root table size t determined by NextExp2_32()

v1.6.2
    2014-11-06
        * RegCluster root table size t is now next power of 2 
        * RegCluster.MembersByID checked before adding member       SLOC 4342
v1.6.1
    2014-11-04
        * switch to 20 byte content keys
v1.6.0
    2014-10-24
        * AdminMember renamed AdminClient
        * MemberInfo Peer becomes pointer; ClusterMember.AddPeers()
            adds other members to Node using AddPeer(), copying     SLOC 4301
        * registry is now a Peer, MemberMaker.RegPeer
        * all cluster members are also Peers to one another         SLOC 4308
v1.5.0
    2014-10-22
        * ClientInfo distinguished from MemberInfo
        * various bugs dealt with; all tests succeed
        * changes to reflect API changes in imported packages       SLOC 4284
v1.4.0
    2014-10-16
        * major changes in API: MemberMaker takes node parameter    SLOC 3990
v1.3.0
    2014-10-15
        * MemberNode() becomes MemberMaker()
        * eph_server_test.go corrected and extended                 SLOC 3990
v1.2.0
    2014-10-07
        * replaced IDMap with HAMT                                  SLOC 3938
v1.1.5
    2014-10-06
        * changed default port to 56789 (45678 for testing)         SLOC 3909
v1.1.4
    2014-09-23
        * dropped AdminClient returning error                       SLOC 3904
v1.1.3
    2014-09-22
        * dropped SoloClient returning error                        SLOC 3906
v1.1.2
    2014-09-20
        * dropped MemberNode.Err
        * MemberNode.DoneCh becomes chan error                      SLOC 3908
v1.1.1
    2014-09-17
        * corrected handling of nil member, client_node.go, 
            GetAndMembers()                                         SLOC 3916
v1.1.0
    2014-09-16
        * MemberNode.JoinAndReply() joins by cluster ID, not name
        * renamed fields, changed some from private to public       SLOC 3914
v1.0.0
    2014-09-15

v0.5.2
    2014-09-14
        * clusterSize, epCount uint32 to match p.proto              SLOC 3903
v0.5.2
    2014-09-13

p.pb.go
// Code generated by protoc-gen-go.
// source: p.proto
// DO NOT EDIT!

/*
Package reg is a generated protocol buffer package.

It is generated from these files:
	p.proto

It has these top-level messages:
	XLRegMsg
*/
package reg

import proto "github.com/golang/protobuf/proto"
import math "math"

// Reference imports to suppress errors if they are not otherwise used.
var _ = proto.Marshal
var _ = math.Inf

type XLRegMsg_Tag int32

const (
	XLRegMsg_RegCredRequest XLRegMsg_Tag = 1
	XLRegMsg_RegCredReply   XLRegMsg_Tag = 2
	XLRegMsg_Member         XLRegMsg_Tag = 3
	XLRegMsg_MemberOK       XLRegMsg_Tag = 4
	// size means cluster size, constrained to 1 < n <= 64
	// epCount is number of menber endPoints, which must be 1 or 2
	XLRegMsg_Create      XLRegMsg_Tag = 5
	XLRegMsg_CreateReply XLRegMsg_Tag = 6
	// Response to Join is clusterID + size + epCount -OR- error
	XLRegMsg_Join       XLRegMsg_Tag = 7
	XLRegMsg_JoinReply  XLRegMsg_Tag = 8
	XLRegMsg_GetCluster XLRegMsg_Tag = 9
	// Response to Get is a list of known members -OR- error
	XLRegMsg_ClusterMembers XLRegMsg_Tag = 10
	XLRegMsg_Bye            XLRegMsg_Tag = 11
	XLRegMsg_Ack            XLRegMsg_Tag = 12
	XLRegMsg_Error          XLRegMsg_Tag = 13
)

var XLRegMsg_Tag_name = map[int32]string{
	1:  "RegCredRequest",
	2:  "RegCredReply",
	3:  "Member",
	4:  "MemberOK",
	5:  "Create",
	6:  "CreateReply",
	7:  "Join",
	8:  "JoinReply",
	9:  "GetCluster",
	10: "ClusterMembers",
	11: "Bye",
	12: "Ack",
	13: "Error",
}
var XLRegMsg_Tag_value = map[string]int32{
	"RegCredRequest": 1,
	"RegCredReply":   2,
	"Member":         3,
	"MemberOK":       4,
	"Create":         5,
	"CreateReply":    6,
	"Join":           7,
	"JoinReply":      8,
	"GetCluster":     9,
	"ClusterMembers": 10,
	"Bye":            11,
	"Ack":            12,
	"Error":          13,
}

func (x XLRegMsg_Tag) Enum() *XLRegMsg_Tag {
	p := new(XLRegMsg_Tag)
	*p = x
	return p
}
func (x XLRegMsg_Tag) String() string {
	return proto.EnumName(XLRegMsg_Tag_name, int32(x))
}
func (x *XLRegMsg_Tag) UnmarshalJSON(data []byte) error {
	value, err := proto.UnmarshalJSONEnum(XLRegMsg_Tag_value, data, "XLRegMsg_Tag")
	if err != nil {
		return err
	}
	*x = XLRegMsg_Tag(value)
	return nil
}

type XLRegMsg struct {
	Op               *XLRegMsg_Tag     `protobuf:"varint,1,opt,enum=reg.XLRegMsg_Tag" json:"Op,omitempty"`
	AesKey           []byte            `protobuf:"bytes,3,opt" json:"AesKey,omitempty"`
	Salt1            []byte            `protobuf:"bytes,4,opt" json:"Salt1,omitempty"`
	Salt2            []byte            `protobuf:"bytes,5,opt" json:"Salt2,omitempty"`
	Version          *uint32           `protobuf:"varint,6,opt" json:"Version,omitempty"`
	MemberName       *string           `protobuf:"bytes,8,opt" json:"MemberName,omitempty"`
	MemberID         []byte            `protobuf:"bytes,9,opt" json:"MemberID,omitempty"`
	MemberAttrs      *uint64           `protobuf:"varint,10,opt" json:"MemberAttrs,omitempty"`
	MemberSpecs      *XLRegMsg_Token   `protobuf:"bytes,11,opt" json:"MemberSpecs,omitempty"`
	ClusterID        []byte            `protobuf:"bytes,13,opt" json:"ClusterID,omitempty"`
	ClusterName      *string           `protobuf:"bytes,14,opt" json:"ClusterName,omitempty"`
	ClusterMaxSize   *uint32           `protobuf:"varint,15,opt" json:"ClusterMaxSize,omitempty"`
	ClusterAttrs     *uint64           `protobuf:"varint,16,opt" json:"ClusterAttrs,omitempty"`
	EndPointCount    *uint32           `protobuf:"varint,17,opt" json:"EndPointCount,omitempty"`
	Which            *uint64           `protobuf:"varint,20,opt" json:"Which,omitempty"`
	Tokens           []*XLRegMsg_Token `protobuf:"bytes,21,rep" json:"Tokens,omitempty"`
	DigSig           []byte            `protobuf:"bytes,22,opt" json:"DigSig,omitempty"`
	ErrDesc          *string           `protobuf:"bytes,23,opt" json:"ErrDesc,omitempty"`
	XXX_unrecognized []byte            `json:"-"`
}

func (m *XLRegMsg) Reset()         { *m = XLRegMsg{} }
func (m *XLRegMsg) String() string { return proto.CompactTextString(m) }
func (*XLRegMsg) ProtoMessage()    {}

func (m *XLRegMsg) GetOp() XLRegMsg_Tag {
	if m != nil && m.Op != nil {
		return *m.Op
	}
	return XLRegMsg_RegCredRequest
}

func (m *XLRegMsg) GetAesKey() []byte {
	if m != nil {
		return m.AesKey
	}
	return nil
}

func (m *XLRegMsg) GetSalt1() []byte {
	if m != nil {
		return m.Salt1
	}
	return nil
}

func (m *XLRegMsg) GetSalt2() []byte {
	if m != nil {
		return m.Salt2
	}
	return nil
}

func (m *XLRegMsg) GetVersion() uint32 {
	if m != nil && m.Version != nil {
		return *m.Version
	}
	return 0
}

func (m *XLRegMsg) GetMemberName() string {
	if m != nil && m.MemberName != nil {
		return *m.MemberName
	}
	return ""
}

func (m *XLRegMsg) GetMemberID() []byte {
	if m != nil {
		return m.MemberID
	}
	return nil
}

func (m *XLRegMsg) GetMemberAttrs() uint64 {
	if m != nil && m.MemberAttrs != nil {
		return *m.MemberAttrs
	}
	return 0
}

func (m *XLRegMsg) GetMemberSpecs() *XLRegMsg_Token {
	if m != nil {
		return m.MemberSpecs
	}
	return nil
}

func (m *XLRegMsg) GetClusterID() []byte {
	if m != nil {
		return m.ClusterID
	}
	return nil
}

func (m *XLRegMsg) GetClusterName() string {
	if m != nil && m.ClusterName != nil {
		return *m.ClusterName
	}
	return ""
}

func (m *XLRegMsg) GetClusterMaxSize() uint32 {
	if m != nil && m.ClusterMaxSize != nil {
		return *m.ClusterMaxSize
	}
	return 0
}

func (m *XLRegMsg) GetClusterAttrs() uint64 {
	if m != nil && m.ClusterAttrs != nil {
		return *m.ClusterAttrs
	}
	return 0
}

func (m *XLRegMsg) GetEndPointCount() uint32 {
	if m != nil && m.EndPointCount != nil {
		return *m.EndPointCount
	}
	return 0
}

func (m *XLRegMsg) GetWhich() uint64 {
	if m != nil && m.Which != nil {
		return *m.Which
	}
	return 0
}

func (m *XLRegMsg) GetTokens() []*XLRegMsg_Token {
	if m != nil {
		return m.Tokens
	}
	return nil
}

func (m *XLRegMsg) GetDigSig() []byte {
	if m != nil {
		return m.DigSig
	}
	return nil
}

func (m *XLRegMsg) GetErrDesc() string {
	if m != nil && m.ErrDesc != nil {
		return *m.ErrDesc
	}
	return ""
}

// the Token describes a member
type XLRegMsg_Token struct {
	Name     *string `protobuf:"bytes,1,opt" json:"Name,omitempty"`
	Attrs    *uint64 `protobuf:"varint,2,opt" json:"Attrs,omitempty"`
	ID       []byte  `protobuf:"bytes,3,opt" json:"ID,omitempty"`
	CommsKey []byte  `protobuf:"bytes,4,opt" json:"CommsKey,omitempty"`
	SigKey   []byte  `protobuf:"bytes,5,opt" json:"SigKey,omitempty"`
	// by convention, MyEnds[0] for inter-cluster comms,
	// MyEnds[1] for cluster-client comms
	// there must be epCount endPoints present
	MyEnds           []string `protobuf:"bytes,6,rep" json:"MyEnds,omitempty"`
	DigSig           []byte   `protobuf:"bytes,7,opt" json:"DigSig,omitempty"`
	XXX_unrecognized []byte   `json:"-"`
}

func (m *XLRegMsg_Token) Reset()         { *m = XLRegMsg_Token{} }
func (m *XLRegMsg_Token) String() string { return proto.CompactTextString(m) }
func (*XLRegMsg_Token) ProtoMessage()    {}

func (m *XLRegMsg_Token) GetName() string {
	if m != nil && m.Name != nil {
		return *m.Name
	}
	return ""
}

func (m *XLRegMsg_Token) GetAttrs() uint64 {
	if m != nil && m.Attrs != nil {
		return *m.Attrs
	}
	return 0
}

func (m *XLRegMsg_Token) GetID() []byte {
	if m != nil {
		return m.ID
	}
	return nil
}

func (m *XLRegMsg_Token) GetCommsKey() []byte {
	if m != nil {
		return m.CommsKey
	}
	return nil
}

func (m *XLRegMsg_Token) GetSigKey() []byte {
	if m != nil {
		return m.SigKey
	}
	return nil
}

func (m *XLRegMsg_Token) GetMyEnds() []string {
	if m != nil {
		return m.MyEnds
	}
	return nil
}

func (m *XLRegMsg_Token) GetDigSig() []byte {
	if m != nil {
		return m.DigSig
	}
	return nil
}

func init() {
	proto.RegisterEnum("reg.XLRegMsg_Tag", XLRegMsg_Tag_name, XLRegMsg_Tag_value)
}
